### PR TITLE
Mask secret value passed between extension scripts

### PR DESCRIPTION
**Pull request by everdimension**

### CODE REVIEW

File: src/ui/shared/requests/useWalletSecretQuery.ts [hunk 1]
```diff
@@ -10,18 +11,22 @@ async function getSecretValue({
   address?: string | null;
   groupId?: string | null;
   seedType: SeedType;
-}) {
+}): Promise<
+  | { seedType: SeedType.privateKey; value: string }
```
Comment (vyorkin) [inline src/ui/shared/requests/useWalletSecretQuery.ts:15]: Could you elaborate a bit on why the `privateKey` is not `LocallyEncoded`?

File: src/shared/wallet/encode-locally.test.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+describe.only('encode-locally.ts', () => {
+  test('encodeForMasking', () => {
+    const value = 'hello';
+    expect(decodeMasked(encodeForMasking(value))).toBe(value);
+  });
+
+  test('maskedValue is not equal to input value', () => {
+    const value = 'something';
+    expect(encodeForMasking(value)).not.toBe(value);
```
Comment (zerts) [inline src/shared/wallet/encode-locally.test.ts:11]: May be it is slightly more reliable to check that encoded value does not include initial value?

File: src/background/Wallet/WalletRecord.ts [hunk 3]
```diff
@@ -505,7 +510,8 @@ export class WalletRecordModel {
     if (!encryptedPhrase) {
       throw new Error('Pending wallet does not have a seed phrase');
     }
-    return decryptMnemonic(encryptedPhrase, credentials);
+    const phrase = await decryptMnemonic(encryptedPhrase, credentials);
+    return encodeForMasking(phrase);
   }
 
   static async getPrivateKey(
```
[the remark this answers is not in the document]
Comment (everdimension) [reply src/background/Wallet/WalletRecord.ts:517]: Yep! I introduced `Opaque` type and it became much easier to track privateKey params as well

File: src/shared/wallet/encode-locally.test.ts (hunk 2 above)
Comment (everdimension) [reply src/shared/wallet/encode-locally.test.ts:11]: why not!

File: src/ui/shared/requests/useWalletSecretQuery.ts (hunk 1 above)
Comment (everdimension) [reply src/ui/shared/requests/useWalletSecretQuery.ts]: It does, updated :)

**Review by zerts — APPROVED** [review]

File: src/shared/type-utils/Opaque.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// eslint-disable-next-line @typescript-eslint/no-namespace
+namespace OpaqueSymbols {
+  export declare const type: unique symbol;
+  export declare const name: unique symbol;
+}
+
+export type Opaque<T, Name> = {
+  readonly [OpaqueSymbols.type]: T;
+  readonly [OpaqueSymbols.name]: Name;
+};
```
Comment (vyorkin) [inline src/shared/type-utils/Opaque.ts:10]: 🔥

File: src/shared/wallet/encode-locally.test.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    expect(decodeMasked(encodeForMasking(value))).toBe(value);
+  });
+
+  test('maskedValue is not equal to input value', () => {
+    const value = 'something';
+    const encoded = encodeForMasking(value);
+    expect(encoded).not.toBe(value);
+  });
+
+  test('maskedValue and value are not substrings of each other', () => {
```
Comment (vyorkin) [inline src/shared/wallet/encode-locally.test.ts:16]: this test actually makes sense!

File: src/background/Wallet/helpers/mask.ts [hunk 6]
```diff
@@ -9,7 +9,10 @@ function maskMnemonic(
   return mnemonic ? { phrase: '<phrase>', path: mnemonic.path } : null;
 }
 
-export function maskWallet(wallet: ExternallyOwnedAccount | BareWallet) {
+/** TODO: rename to "removeSensitiveValues"? or something better? */
+export function maskWallet<T extends ExternallyOwnedAccount | BareWallet>(
```
Comment (vyorkin) [inline src/background/Wallet/helpers/mask.ts:13]: `maskSensitiveData` :)

**Review by vyorkin — APPROVED** [review]

File: src/background/Wallet/helpers/mask.ts (hunk 6 above)
Comment (everdimension) [reply src/background/Wallet/helpers/mask.ts:13]: Well, none of these fit as this helper basically removes the value by replacing it with a fixed string

**Pull request merged**